crates/sui-core/src/execution_scheduler/balance_withdraw_scheduler/scheduler.rs

```diff
@@ -0,0 +1,156 @@
+        let (withdraw_sender, withdraw_receiver) =
+            unbounded_channel("withdraw_scheduler_withdraws");
+        let (settlement_sender, settlement_receiver) =
+            unbounded_channel("withdraw_scheduler_settlements");
+        let scheduler = Arc::new(Self {
+            inner,
+            withdraw_sender,
+            settlement_sender,
+        });
+        tokio::spawn(scheduler.clone().process_withdraw_task(withdraw_receiver));
+        tokio::spawn(
+            scheduler
+                .clone()
+                .process_settlement_task(settlement_receiver, init_accumulator_version),
+        );
+        scheduler
+    }
+
+    /// This function will be called either by ConsensusHandler or the CheckpointExecutor.
+    /// It will be called at most once per consensus commit batch that all reads the same root accumulator version.
+    /// If a consensus commit batch does not contain any withdraw reservations, it can skip calling this function.
+    /// It must be called sequentially in order to correctly schedule withdraws.
+    /// It is OK to call this function multiple times for the same accumulator version (which will happen between
+    /// the calls to the function by ConsensusHandler and the CheckpointExecutor).
+    #[allow(dead_code)]
+    pub fn schedule_withdraws(
+        &self,
+        accumulator_version: SequenceNumber,
+        withdraws: Vec<TxBalanceWithdraw>,
+    ) -> BTreeMap<TransactionDigest, oneshot::Receiver<ScheduleResult>> {
+        let (reservations, receivers) = WithdrawReservations::new(accumulator_version, withdraws);
+        if let Err(err) = self.withdraw_sender.send(reservations) {
+            tracing::error!("Failed to send withdraw reservations: {:?}", err);
+        }
+        receivers
+    }
+
+    /// This function is called whenever a new version of the accumulator root object is committed
+    /// in the writeback cache.
+    /// It is OK to call this function out of order, as the implementation will handle the out of order calls.
+    /// It must be called once for each version of the accumulator root object.
+    #[allow(dead_code)]
+    pub fn settle_balances(&self, settlement: BalanceSettlement) {
+        if let Err(err) = self.settlement_sender.send(settlement) {
+            tracing::error!("Failed to send balance settlement: {:?}", err);
+        }
+    }
+
+    async fn process_withdraw_task(
+        self: Arc<Self>,
+        mut withdraw_receiver: UnboundedReceiver<WithdrawReservations>,
+    ) {
+        let mut last_scheduled_version = None;
+        while let Some(event) = withdraw_receiver.recv().await {
+            if let Some(last_scheduled_version) = last_scheduled_version {
+                if event.accumulator_version <= last_scheduled_version {
+                    // It is possible to receive withdraw reservations for the same accumulator version
+                    // multiple times due to the race between consensus and checkpoint execution.
```

Comment:
this problem goes away if we only accept reservations from consensus

Reply:
This happens because it can be called from both consensus and checkpoint execution. They would both be in order, but one can be ahead of another.